### PR TITLE
Feature/find nearest store

### PR DESCRIPTION
**Pull request by phillipet**

### CODE REVIEW

File: server/schema/order/res.js [hunk 1]
```diff
@@ -39,6 +42,39 @@ function getModifiersPrice(modifiers, modifierIds) {
     throw new Error(error.message)
   }
 }
+async function findNearestStore(deliveryAddress) {
+  let stores = await Store.findAll()
+  let storeAddresses = getStoreAddresses(stores)
+  let url =
+    'https://maps.googleapis.com/maps/api/distancematrix/json?origins=' + deliveryAddress + '&destinations=' + storeAddresses + '&key=' + apiKey
+  let distances = await fetch(url)
+    .then(res => res.json())
```
Comment (hemvanh) [inline server/schema/order/res.js]: .then(async res=> await res.json())

File: server/schema/order/res.js [hunk 2]
```diff
@@ -39,6 +42,39 @@ function getModifiersPrice(modifiers, modifierIds) {
...
+      return json
+    })
+  let storeName = findNearestStoreName(distances)
+  return stores.find(store => store.get('addressGmap') === storeName).get('id')
+}
+function findNearestStoreName(distances) {
+  let elements = distances.rows[0].elements,
+    minElement = distances.rows[0].elements[0].distance.value,
+    minIndex = 0
+  for (let i = 1; i < elements.length; i++) {
```
Comment (hemvanh) [inline server/schema/order/res.js]: let minIndex  = _d.findIndex(elements ,_d.min(elements,'distance.value' ))
... or could try something cleaner

File: server/migrations/20190110045634-create-stores.js [hunk 3]
```diff
@@ -14,6 +14,9 @@ module.exports = {
       address: {
         type: Sequelize.STRING(500),
       },
+      addressGmap: {
```
Comment (hemvanh) [inline server/migrations/20190110045634-create-stores.js]: gmapAddress
--> we follow Wnglish grammar

**Review by hemvanh — CHANGES_REQUESTED** [review]

File: server/schema/order/res.js [hunk 4]
```diff
@@ -39,6 +43,28 @@ function getModifiersPrice(modifiers, modifierIds) {
...
+  return stores.find(store => store.get('gmapAddress') === storeName).get('id')
+}
+function findNearestStoreName(distances) {
+  let elements = _d.map(distances.rows[0].elements, 'distance.value')
+  let minIndex = elements.indexOf(_d.min(elements))
+  return distances.destination_addresses[minIndex]
+}
+function getStoreAddresses(stores) {
+  let strAddress = ''
+  stores.forEach(store => {
```
Comment (hemvanh) [inline server/schema/order/res.js]: _d.map(stores,'gmapAddress').join('|')

File: server/schema/order/res.js [hunk 5]
```diff
@@ -39,6 +43,28 @@ function getModifiersPrice(modifiers, modifierIds) {
     throw new Error(error.message)
   }
 }
+async function findNearestStore(deliveryAddress) {
```
Comment (hemvanh) [inline server/schema/order/res.js]: findNearestStore  -> findNearestStoreId

File: server/schema/order/res.js [hunk 6]
```diff
@@ -39,6 +43,28 @@ function getModifiersPrice(modifiers, modifierIds) {
...
+  let stores = await Store.findAll()
+  let storeAddresses = getStoreAddresses(stores)
+  let url =
+    'https://maps.googleapis.com/maps/api/distancematrix/json?origins=' + deliveryAddress + '&destinations=' + storeAddresses + '&key=' + apiKey
+  let distances = await fetch(url).then(async res => await res.json())
+  let storeName = findNearestStoreName(distances)
+  return stores.find(store => store.get('gmapAddress') === storeName).get('id')
```
Comment (hemvanh) [inline server/schema/order/res.js]: _d.find(stores,{gmapAddress:storeName}).get('id')

**Review by hemvanh — CHANGES_REQUESTED** [review]

File: server/schema/order/res.js (hunk 4 above)
Comment (phillipet) [reply server/schema/order/res.js]: stores is a sequelize object, can not map it http://prntscr.com/m9694n

File: server/schema/order/res.js (hunk 6 above)
Comment (phillipet) [reply server/schema/order/res.js]: stores is a sequelize object, can not map it http://prntscr.com/m9694n

Comment (phillipet) [reply server/schema/order/res.js]: Additionally, using map then find will loop array twice times(performance isn't good)

Comment (hemvanh) [reply server/schema/order/res.js]: _d.find(stores,{dataValues:{gmapAddress:storeName}}).get('id')

File: server/schema/order/res.js (hunk 4 above)
Comment (hemvanh) [reply server/schema/order/res.js]: _d.map(stores,'dataValues.gmapAddress').join('|')

**Review by hemvanh — APPROVED** [review]

**Pull request merged**